devel/boost: make boost-numpy extension a subport…

mamoll opened the pull request:
#### Description
…  to reduce the large number of dependencies that are being pulled in. It also fixes the circular dependency of boost on itself. See PR #1309 and https://trac.macports.org/ticket/55807.

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->
<!-- (delete all below for minor changes) -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.13.3 17D102
Xcode 9.2 9C40b 

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

michaelld commented:
Looks good to me. @ryandesign your take?

ryandesign commented:
I haven't tried it. If it works for you, you can merge it.

mamoll commented:
Agreed. I have pushed another commit that does this.

michaelld commented:
Yes good point; thanks for handling so promptly.

A few minor tweaks, if you choose to do them:
* you don't need to specify "revision 0" ... that's implied.
* there is some end-of-line whitespace that I prefer to not have (I'll clean it up sometime down the road if need be).
Neither of these is a deal-breaker.

ryandesign commented:
> you don't need to specify "revision 0" ... that's implied.

I recommend specifying `revision 0` in ports that have subports. It makes it clear to others wanting to increase a port's revision that they have to consider which of the several ports in the portfile need their revisions increased.

michaelld commented:
OK ... that makes sense. I'm good either way. I'll test the latest changes out shortly.

michaelld commented:
Should `${prefix}/include/boost/python/numpy*` be included in the `boost-numpy` port instead of `boost`? guessing so, but I'm not 100% sure.

ryandesign commented:
I haven't a clue.

mamoll commented:
Yes, probably. Right now it works, because boost-bumpy depends on boost. The header `${prefix}/include/boost/python/numpy*` files in boost are not really useful without the boost-numpy library. Let me fix that.

michaelld commented:
Poking through the boost/numpy/* headers, it looks like the answer is "yes"; these headers are the API to the boost_numpy library. So, @mamoll please tweak accordingly.

mamoll commented:
There are also numpy-specific documentation pages and code samples. I can move those to the boost-numpy subport as well, but it'll break some links (e.g., in `${prefix}/share/doc/boost/libs/python/doc/html/index.html`).

michaelld commented:
I'm for keeping the API and ABI together; don't care about the docs so much.

mamoll commented:
Ok, moved just the header files. I reinstalled boost and boost-numpy, and verified that the headers are now in the right port.

michaelld commented:
Great. I testing right now on 10.8 and 10.12.

michaelld commented:
OK looks good to me now. Tested on 10.8, 10.12, and 10.13. All build / install / contents same-same.

mamoll commented:
Is there anything else I need to do, or is this essentially ready to get merged?

michaelld commented:
I'm good with the merge. @ryandesign ?

ryandesign commented:
Why does the subport `delete ${destroot}${prefix}/Library`? Did you mean to `delete ${destroot}${frameworks_dir}` instead? Is there actually something in there that you need to delete? When I tried destrooting, and exiting before this step, ${destroot}${frameworks_dir} existed but was empty.

A `post-destroot` block was added to the main port which deletes the numpy includes but which also installs documentation. The port already installed some documentation without this block, so now some documentation is installed twice. The examples, for example, are installed both in /opt/local/share/doc/boost/libs/python/example and in /opt/local/share/doc/boost/libs/python/example/example. I don't see how this relates to the matter of the boost-python subport and I think the parts of the block that install documentation should be removed.

All of boost's variants are made available to the boost-numpy subport too. I see why we want no_single, no_static, and the compiler and python variants, but what about regex_match_extra?

The subport verifies that its python variant matches boost's python variant, but shouldn't it also verify that the no_single, no_static and compiler variants match too?

mamoll commented:
> Why does the subport delete ${destroot}${prefix}/Library? Did you mean to delete ${destroot}${frameworks_dir} instead? Is there actually something in there that you need to delete? 

Fixed in the latest commit. If an mpi and a python variant are active, a python module gets installed, which needs to be removed for the boost-numpy subport.

> The port already installed some documentation without this block, so now some documentation is installed twice. 

This problem already existed before this PR, but I have removed the unnecessary code.

> All of boost's variants are made available to the boost-numpy subport too. I see why we want no_single, no_static, and the compiler and python variants, but what about regex_match_extra?

Fixed.

> The subport verifies that its python variant matches boost's python variant, but shouldn't it also verify that the no_single, no_static and compiler variants match too?

I am not sure whether all compiler variants need to match, but I have added checks to make sure the no_single and no_static variants match.

ryandesign commented:
>> The port already installed some documentation without this block, so now some documentation is installed twice.
> 
> This problem already existed before this PR, but I have removed the unnecessary code.

Thanks. I thought it was [this code that you added in this PR](https://github.com/macports/macports-ports/pull/1401/commits/9291dcd30640b47ccea562d7df1719fb866db4c5?w=1#diff-a879913c9e32da575ee3081a0b25a791R284).

mojca commented on devel/boost/Portfile:
```diff
@@ -251,51 +253,97 @@ variant no_single description {Disable building single-threaded libraries} {
+        }
+    }
+    if {[variant_isset no_static]} {
+        require_active_variants boost no_single
```
Should this be `no_static`?

mamoll replied:
Yes. Fixed in the latest commit.

pmetzger commented:
@mamoll I believe I can now merge this if you could squash the commits.

mamoll commented:
@pmetzger, how do I do this? I tried this:
```
git reset --soft 9aef396e4c29007abf037434284b5e3858bfafe3
git commit -m "devel/boost: make boost-numpy extension a subport to reduce the large number of dependencies that are being pulled in"
git push mamoll 
To github.com:mamoll/macports-ports.git
 ! [rejected]              boost-numpy-fix -> boost-numpy-fix (non-fast-forward)
error: failed to push some refs to 'git@github.com:mamoll/macports-ports.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
I am not enough of a git expert to figure out how to fix this.

mojca commented:
`git push -f mamoll`

mamoll commented:
@mojca, that worked. Thanks!

pmetzger commented:
Merged. @mamoll, thank you for your patience.

michaelld commented:
Yes! and perseverance! alls well that ends well ;)
